pkg/receive/handler.go

```diff
@@ -479,5 +504,47 @@ func isConflict(err error) bool {
 	if err == nil {
 		return false
 	}
-	return err == conflictErr || err == storage.ErrDuplicateSampleForTimestamp || err == storage.ErrOutOfOrderSample || err == storage.ErrOutOfBounds || err.Error() == strconv.Itoa(http.StatusConflict)
+	return err == conflictErr ||
+		err == storage.ErrDuplicateSampleForTimestamp ||
+		err == storage.ErrOutOfOrderSample ||
+		err == storage.ErrOutOfBounds ||
+		err.Error() == strconv.Itoa(http.StatusConflict) ||
+		status.Code(err) == codes.AlreadyExists
+}
+
+func newPeerGroup(dialOpts ...grpc.DialOption) *peerGroup {
+	return &peerGroup{
+		dialOpts: dialOpts,
+		cache:    map[string]storepb.WriteableStoreClient{},
+		m:        sync.Mutex{},
+		dialer:   grpc.DialContext,
+	}
+}
+
+type peerGroup struct {
+	dialOpts []grpc.DialOption
+	cache    map[string]storepb.WriteableStoreClient
+	m        sync.Mutex
```

Comment:
if we are making thousands of requests a second, then pretty soon every node in the hashring will have already messaged every other node. Since we won't be writing so often to the cache, I think we  could benefit greatly from using a RW mutex here